### PR TITLE
Mvc\Model::__get() / __set() and related storage

### PR DESCRIPTION
**Pull request by zsilbi**
Hello!

* Type: bug fix
* Link to issues: https://github.com/phalcon/cphalcon/issues/13531, https://github.com/phalcon/cphalcon/pull/13616, https://github.com/phalcon/cphalcon/issues/13964, https://github.com/phalcon/cphalcon/issues/14021

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [x] I updated the CHANGELOG

Added new related storages
Modified __set(), __get(), getRelated() to use these related storages
Modified save() to use these related storages with transaction handling
Modified _getRelatedRecords() to use getRelated() when possible
Reusable relations are cached and returned from the Model directly
Non-reusable relations are not cached
Fixed __set() to handle arrays correctly

Added tests for Mvc\Model::save()
Added tests for Mvc\Model::__set()
Added tests for Mvc\Model::writeAttribute()
More tests needed...

Thanks,
zsilbi

### CODE REVIEW

File: tests/integration/Mvc/Model/UnderscoreSetCest.php [hunk 1]
```diff
@@ -12,20 +12,301 @@
...
+    {
+        $I->wantToTest('Tests Mvc\Model - __set() with has-one related record');
+
+        $customer = new Models\Customers();
+        $customer->user = new Models\Users();
+
+        $user = $customer->user;
+
+        $I->assertInstanceOf(Models\Users::class, $user);
+        $I->assertEquals($customer->getDirtyState(), Model::DIRTY_STATE_TRANSIENT);
```
Comment (SidRoberts) [inline tests/integration/Mvc/Model/UnderscoreSetCest.php]: Assertions are `($expected, $actual)` so it should be:

```php
$I->assertEquals(
    Model::DIRTY_STATE_TRANSIENT,
    $customer->getDirtyState()
);
```

And space out the assertions on multiple lines to make it easier to read. :wink:

File: tests/integration/Mvc/Model/UnderscoreSetCest.php [hunk 2]
```diff
@@ -12,20 +12,301 @@
...
+
+        $robot = new Models\Robots();
+        $robot->robotsParts = [
+            new Models\RobotsParts(),
+            new Models\RobotsParts()
+        ];
+
+        $robotsParts = $robot->robotsParts;
+
+        $I->assertTrue(is_array($robotsParts));
```
Comment (SidRoberts) [inline tests/integration/Mvc/Model/UnderscoreSetCest.php]: https://phpunit.readthedocs.io/en/7.4/assertions.html#assertinternaltype

File: tests/integration/Mvc/Model/UnderscoreSetCest.php (hunk 1 above)
Comment (zsilbi) [reply tests/integration/Mvc/Model/UnderscoreSetCest.php]: Thank you, @SidRoberts :)

File: tests/integration/Mvc/Model/UnderscoreSetCest.php (hunk 2 above)
Comment (zsilbi) [reply tests/integration/Mvc/Model/UnderscoreSetCest.php]: Resolved, thanks!

**Comment by zsilbi** [comment]
Is caching related records in `Model` worth it?
`Manager` handles it well, and I don't think it has any performance benefits to do the same here. 

Making `isRelationshipLoaded()` work would require this cache, but I think `Manager` could be used for this kind of tasks, but I honestly don't see any real-world use case for this function from my perspective.

Also, records queried with any extra arguments, like
```php
$mechanicalParts = $robot->getRelated('parts', ['type = "mechanical"');
```
will only be cached in `Manager`.

**Comment by CameronHall** [comment]
@zsilbi I'm a little naive to SplObjectStorage and observing when there's a use case for it, but would that serve as a better option than storing the relationship objects in an array? Totally chill if you think otherwise, I'm just curious.

**Comment by zsilbi** [comment]
@CameronHall: I checked SplObjectStorage but I dont't think it would be beneficial to use here.
The related records are indexed by their aliases, not by their content (hash).
On the other hand, I think in `Manager`, if you have to store any data for a given object, it could work.

**Comment by niden** [comment]
@zsilbi can you please rebase this?

This looks great. Awesome work!

**Comment by niden** [comment]
Thank you @zsilbi

**Pull request merged**